vyper/ir/compile_ir.py

```diff
@@ -945,41 +966,55 @@ def _prune_unused_jumpdests(assembly):
     i = 0
     while i < len(assembly) - 2:
         if is_symbol(assembly[i]) and assembly[i] not in used_jumpdests:
+            changed = True
             del assembly[i : i + 2]
         else:
             i += 1
 
+    return changed
+
 
 def _stack_peephole_opts(assembly):
+    changed = False
     i = 0
     while i < len(assembly) - 2:
         # usually generated by with statements that return their input like
         # (with x (...x))
         if assembly[i : i + 3] == ["DUP1", "SWAP1", "POP"]:
             # DUP1 SWAP1 POP == no-op
+            changed = True
             del assembly[i : i + 3]
             continue
         # usually generated by nested with statements that don't return like
         # (with x (with y ...))
         if assembly[i : i + 3] == ["SWAP1", "POP", "POP"]:
             # SWAP1 POP POP == POP POP
+            changed = True
             del assembly[i]
             continue
         i += 1
 
+    return changed
+
 
 # optimize assembly, in place
 def _optimize_assembly(assembly):
     for x in assembly:
         if isinstance(x, list):
             _optimize_assembly(x)
 
-    _prune_unreachable_code(assembly)
-    _merge_iszero(assembly)
-    _merge_jumpdests(assembly)
-    _prune_inefficient_jumps(assembly)
-    _prune_unused_jumpdests(assembly)
-    _stack_peephole_opts(assembly)
+    while True:
```

Comment:
Should probably have an explicit bounds check of number of cycles, or else you could probably find an infinite loop somewhere and get stuck

Reply:
infinite loop would indicate a bug, since if one of the steps does not change the assembly it should report `changed = False`

Reply:
anyways -- addressed in https://github.com/vyperlang/vyper/pull/2837/commits/935b77b68f1a7d918ab64857ff66913508668a7e